src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs

```diff
@@ -1,47 +1,33 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Buffers;
 using System.Collections;
 using System.Diagnostics;
 using System.Runtime.InteropServices;
+using System.Text;
 
 namespace System
 {
     public static partial class Environment
     {
         private static string? GetEnvironmentVariableCore(string variable)
         {
-            Span<char> buffer = stackalloc char[128]; // a somewhat reasonable default size
-            int requiredSize = Interop.Kernel32.GetEnvironmentVariable(variable, buffer);
+            var builder = new ValueStringBuilder(stackalloc char[128]);
 
-            if (requiredSize == 0 && Marshal.GetLastWin32Error() == Interop.Errors.ERROR_ENVVAR_NOT_FOUND)
+            uint length;
+            while ((length = Interop.Kernel32.GetEnvironmentVariable(variable, ref builder.GetPinnableReference(), (uint)builder.Capacity)) > builder.Capacity)
             {
-                return null;
+                builder.EnsureCapacity((int)length);
```

Comment:
I don't think this is possible, but in theory if `(int)length` evaluates to a negative number then `EnsureCapacity` will no-op and this loop will run infinitely. Do we need a guard against this? (Again, not sure how this could ever happen.)

Reply:
I have added some defensive checks to ValueStringBuilder to throw exception if this ever happens.